File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/BootReceiver.java
package org.gozaltech.nvdaremotecompanion.android;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.util.Log;

public class BootReceiver extends BroadcastReceiver {

    private static final String TAG = "NVDARemote/Boot";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (!Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) return;
        if (!AppPrefs.getAutostart(context)) return;

        Log.i(TAG, "Boot completed — starting ConnectionService");
        Intent svcIntent = new Intent(context, ConnectionService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(svcIntent);
        } else {
            context.startService(svcIntent);
        }
    }
}

File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/service/BootReceiver.kt
package org.gozaltech.nvdaremotecompanion.android.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.gozaltech.nvdaremotecompanion.android.data.SettingsRepository
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class BootReceiver : BroadcastReceiver(), KoinComponent {

    private val settingsRepository: SettingsRepository by inject()

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != Intent.ACTION_BOOT_COMPLETED) return
        val pendingResult = goAsync()
        CoroutineScope(SupervisorJob() + Dispatchers.IO).launch {
            try {
                if (settingsRepository.current().autostart) {
                    context.startForegroundService(ConnectionService.startedFromBackground(context))
                }
            } finally {
                pendingResult.finish()
            }
        }
    }
}
